File: programmers/Courses/알고리즘-문제-해설/7.py
#
# Created by habi on 2023-08-04.
#
# Problem
# https://school.programmers.co.kr/learn/courses/18/lessons/1882
#


def solution(strs, t):
    answer = 0
    N = len(t)

    dp = [0 for i in range(N + 5)]

    for i in range(N):
        dp[i] = 999

    for i in range(N):
        idx = N - i - 1
        word = t[idx:]

        for j in range(len(word)):
            sub_word = word[: (j + 1)]

            if sub_word in strs:
                dp[idx] = min(dp[idx], dp[idx + len(sub_word)] + 1)

    answer = dp[0] if dp[0] != 999 else -1

    return answer


def solution2(strs, t):
    answer = 0
    N = len(t)
    INF = 99999

    dp = [INF for i in range(N + 1)]
    dp[N] = 0

    for i in range(N - 1, -1, -1):
        tmp = ""

        for j in range(5):
            if i + j < N:
                tmp += t[i + j]

                if tmp in strs:
                    # (i + j)로 하나의 조각을 완성하고, 나머지 최소 조각 수는 dp[i + j + 1]에 저장되어 있음
                    dp[i] = min(dp[i], dp[i + j + 1] + 1)

    answer = dp[0] if dp[0] != INF else -1

    return answer

File: programmers/Courses/알고리즘-문제-해설/7.cpp
//
// Created by habi on 2023-08-07.
//
// Problem
// https://school.programmers.co.kr/learn/courses/18/lessons/1882
//

#include <algorithm>
#include <iostream>
#include <set>
#include <string>
#include <vector>
using namespace std;

int solution(vector<string> strs, string t) {
  int answer = 0;
  int N = t.length();
  const int INF = 99999;

  int dp[20001];

  for (int i = 0; i <= N; i++) {
    dp[i] = i < N ? INF : 0;
  }

  for (int i = N - 1; i >= 0; --i) {
    string tmp = "";

    // 단어 조각의 길이는 1 이상 5 이하의 조건을 사용함
    for (int j = 0; j < 5; j++) {
      if (i + j < N) {
        tmp += t[i + j];

        vector<string>::iterator iter = find(strs.begin(), strs.end(), tmp);
        if (iter != strs.end()) {
          // (i + j)로 하나의 조각을 완성하고,
          // 나머지 최소 조각 수는 dp[i + j + 1]에 저장되어 있음
          dp[i] = min(dp[i], dp[i + j + 1] + 1);
        }
      }
    }
  }

  answer = dp[0] != INF ? dp[0] : -1;

  return answer;
}
